### ISSUE TITLE
Unable to cache conversation with FilesystemCache when running on socket

### ISSUE THREAD

**Issue by robertodormepoco** [botman/botman#728]
- BotMan Version: 2.3.0
- PHP Version: 7.0.25
- Messaging Service(s): slack
- Cache Driver: DoctrineCache + FilesystemCache

### Description:
Trying to store a conversation with FilesystemCache having BotMan running with a socket gives "Uncaught Exception: Serialization of 'Closure' is not allowed", due to the explicit logic declared in HandlesConversation::serializeClosure

### Steps To Reproduce:

$botman = BotManFactory::createForSocket(
    [
        'slack' => [
            'token' => getenv('SLACK_API_TOKEN'),
        ],
    ],
    $loop,
    new DoctrineCache(new FilesystemCache(__DIR__))
);

$botman->hears(
    'foo',
    function (\BotMan\BotMan\BotMan $bot) {
        $conversation = new FooConversation();
        $bot->startConversation($conversation);
    }
);

...
React Server listening for https request

BotManFactory::passRequestToSocket(8080, $request);
...

**Comment by mpociot** [comment]
Hi,

yes, that's because of the way that conversations and drivers get stored internally. The easiest fix would be to forbid passing in different cache drivers for the socket usage, as it currently only works with the array cache.
To be honest - the whole socket part is not really fleshed out (and therefore still undocumented).

**Comment by robertodormepoco** [comment]
@mpociot Sorry to bother, but I don't get the part about the reason behind this, the socket server is not coupled for any technical reason with caching, besides the fact that DoctrineCache is not Promise based as React/Cache which i think should be, but anyhow i don't see the limit here.

**Comment by mpociot** [comment]
Yeah, the problem is not related to the caching driver directly, but to how conversations get cached / retrieved right now.

So if you start a conversation, the library serializes the closure which also contains the BotMan instance itself. To prevent serializing the driver instance, we just store the driver name in the serialized BotMan instance.

When you pull the conversation out of the cache to call it again, the BotMan instance `__wakeup` method gets called, which then loads the driver again, based on the name that was serialized.

This works fine if we think of a standard HTTP flow where requests hit your application - and the request defines the driver.
With sockets, this becomes a bit more complicated. We basically do not want to serialize the BotMan instance in this case, because it is available in the long-running process, that has the socket open.

So if we would serialize the BotMan instance for socket connections too, we would not be able to load the driver again, since the driver is basically the socket connection.

**Comment by robertodormepoco** [comment]
Thanks, now it's crystal clear, what about decoupling the BotMan instance from the Conversation instance?
As i see the conversation is always retrieved from a BotMan instance, and there is no driver coupling with the conversation actions, maybe we can inject the BotMan instance from a context of execution, which is not by itself the BotMan instance with the HandlesConversations trait.

And btw, i was looking at the serialized content of the conversation, there is no trace of the BotMan instance

simple test case
```
$request = new Request();
$cache = new DoctrineCache(new FilesystemCache(__DIR__));
$bot = BotManFactory::create([
    [
        'slack' => [
            'token' => ''
        ],
    ],
], $cache, $request);

$conversation = new TestConversation();
$bot->startConversation($conversation);
```

this is what i get

```
1519820449
a:5:{s:12:"conversation";O:47:"RobertoDormePoco\Conversations\TestConversation":2:{s:8:"^@*^@token";N;s:12:"^@*^@cacheTime";N;}s:8:"question";s:482:"O:40:"BotMan\BotMan\Messages\Outgoing\Question":4:{s:10:"^@*^@actions";a:2:{i:0;a:6:{s:4:"name";s:3:"yes";s:4:"text";s:3:"Yes";s:9:"image_url";N;s:4:"type";s:6:"button";s:5:"value";s:3:"yes";s:10:"additional";a:0:{}}i:1;a:6:{s:4:"name";s:2:"no";s:4:"text";s:2:"No";s:9:"image_url";N;s:4:"type";s:6:"button";s:5:"value";s:2:"no";s:10:"additional";a:0:{}}}s:7:"^@*^@text";s:11:"Is foo bar?";s:14:"^@*^@callback_id";s:10:"is_foo_bar";s:11:"^@*^@fallback";s:29:"Unable to fullfil the request";}";s:20:"additionalParameters";s:6:"a:0:{}";s:4:"next";s:296:"C:32:"Opis\Closure\SerializableClosure":250:{a:5:{s:3:"use";N;s:8:"function";s:168:"function (\BotMan\BotMan\Messages\Incoming\Answer $answer) {
            $selectedValue = $answer->getValue();
            $selectedText = $answer->getText();
        }";s:5:"scope";N;s:4:"this";N;s:4:"self";N;}}";s:4:"time";s:21:"0.82115200 1519818649";}
```

**Comment by mpociot** [comment]
Yeah you're right, when a conversation gets loaded, the current BotMan instance will be passed to this conversation: https://github.com/botman/botman/blob/master/src/Traits/HandlesConversations.php#L249

> Maybe we can inject the BotMan instance from a context of execution, which is not by itself the BotMan instance with the HandlesConversations trait.

Yes, that would work. 
I also want to refactor the BotMan class into two or more classes. 
For example right now it holds all methods that are only available once a message is loaded, like:

- getUser
- reply
- randomReply
- startConversation (except when originating conversations)
- ask
- types
- typesAndWaits

I want to split it into two classes - one is responsible for building the commands that the bot listens to - and maybe even allows you to originate messages using a userid + driver.
And one class that will be used in closures and conversations.
